### PR TITLE
HDDS-7349. Flaky integration test have memory leak for RatisDropwizardExports

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

1. Additional RatisDropwizardExports old by DNs after its closed
2. register of RatisDropwizardExports even its already present with name in ratisMetricsMap

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7349

## How was this patch tested?

Tested with flaky and taking heapdump for objects of RatisDropwizardExports referred by DNs ratisMetricsMap.

### CODE REVIEW

**Comment by sumitagrawl** [comment]
@nandakumar131 @ChenSammi Please review

**Comment by sumitagrawl** [comment]
@szetszwo plz review

**Comment by szetszwo** [comment]
@sumitagrawl , each datanode/om/scm creates its own MetricRegistries instead using global(); see https://issues.apache.org/jira/secure/attachment/13051264/MetricRegistries.patch

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java [hunk 1]
```diff
@@ -43,14 +44,15 @@ public RatisDropwizardExports(MetricRegistry registry) {
   }
 
   public static void registerRatisMetricReporters(
-      Map<String, RatisDropwizardExports> ratisMetricsMap) {
+      Map<String, RatisDropwizardExports> ratisMetricsMap,
+      AtomicBoolean isStopped) {
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java:48]: Use `BooleanSupplier`.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java [hunk 2]
```diff
@@ -69,12 +71,19 @@ public static void clear(Map<String, RatisDropwizardExports>
   }
 
   private static void registerDropwizard(RatisMetricRegistry registry,
-      Map<String, RatisDropwizardExports> ratisMetricsMap) {
+      Map<String, RatisDropwizardExports> ratisMetricsMap,
+      AtomicBoolean isStopped) {
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java:75]: Use `BooleanSupplier`.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java [hunk 3]
```diff
@@ -69,12 +71,19 @@ public static void clear(Map<String, RatisDropwizardExports>
...
+    if (isStopped.get()) {
+      return;
+    }
+    
+    String name = registry.getMetricRegistryInfo().getName();
     RatisDropwizardExports rde = new RatisDropwizardExports(
         registry.getDropWizardMetricRegistry());
-    CollectorRegistry.defaultRegistry.register(rde);
-    String name = registry.getMetricRegistryInfo().getName();
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/RatisDropwizardExports.java:76]: Keep `name` declared after `rde`, i.e.
```java
     RatisDropwizardExports rde = new RatisDropwizardExports(
         registry.getDropWizardMetricRegistry());
     String name = registry.getMetricRegistryInfo().getName();
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 4]
```diff
@@ -438,6 +439,7 @@ private enum State {
   private static final int MSECS_PER_MINUTE = 60 * 1000;
 
   private final boolean isSecurityEnabled;
+  private final AtomicBoolean isStopped = new AtomicBoolean(false);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:442]: Use `omState` and do not add `isStopped`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 5]
```diff
@@ -1967,7 +1969,7 @@ private void initializeRatisServer(boolean shouldBootstrap)
       if (omRatisServer == null) {
         // This needs to be done before initializing Ratis.
         RatisDropwizardExports.
-            registerRatisMetricReporters(ratisMetricsMap);
+            registerRatisMetricReporters(ratisMetricsMap, isStopped);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:1972]: Use `() -> omState == State.STOPPED`.

**Review by szetszwo — COMMENTED** [review]
@sumitagrawl , the change looks good.  Some comments inlined; see also https://issues.apache.org/jira/secure/attachment/13051276/3858_review.patch

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
> ...; see also https://issues.apache.org/jira/secure/attachment/13051276/3858_review.patch

@sumitagrawl , just found that I uploaded a wrong file.  It was great that you have fixed everything already!  Anyway, this is the correct file https://issues.apache.org/jira/secure/attachment/13051439/3858_review2.patch

**Comment by adoroszlai** [comment]
Thanks @sumitagrawl for the patch, @szetszwo for the review.

**Pull request merged**